### PR TITLE
GCP-859: scope PSC NAT subnet discovery to the management cluster VPC

### PR DESCRIPTION
**Pull request by cristianoveiga**
## Summary

Fixes PSC NAT subnet discovery to scope subnet lookup to the management cluster's VPC, preventing cross-VPC selection when multiple management clusters share a GCP project.

**Root cause:** \`discoverNATSubnet()\` listed all PSC-purpose subnets in a project/region with no VPC filter. GCP rejects a Service Attachment whose NAT subnet and forwarding rule are in different VPCs, causing PSC setup to fail in multi-cluster-per-project deployments.

**Changes:**

- Derives the management cluster's VPC from the forwarding rule's \`Network\` field (authoritative, set by GCP CCM at ILB creation) and passes it as a server-side filter to \`Subnetworks.List\`
- Renames \`lookupForwardingRuleName\` → \`lookupForwardingRule\` returning \`*compute.ForwardingRule\` to carry both the name and network URL in a single lookup
- Handles the partial-write edge case where \`ForwardingRuleName\` is set but \`NATSubnet\` is not (controller crashed between the two writes)
- Extracts a \`ComputeClient\` interface over the GCP Compute API calls (following the Azure \`PrivateLinkServicesAPI\` pattern), backed by a \`computeServiceAdapter\` that bridges the GCP SDK's chained-call surface. Enables unit testing without GCP credentials

Fixes: https://redhat.atlassian.net/browse/GCP-859

## Test plan

- [ ] Existing unit tests pass (\`go test ./hypershift-operator/controllers/platform/gcp/...\`)
- [ ] New unit tests cover all new code paths via \`fakeGCPComputeClient\`: \`lookupForwardingRule\` (API error, no results, single/multiple results), \`reconcileGCPPrivateServiceConnectSpec\` (early return, lookup error, ILB not provisioned, empty Network field, happy path), \`discoverNATSubnet\` (API error, no subnets, in-use skip, all subnets in use)
- [ ] New e2e test \`[Feature:GCPPrivateServiceConnect]\` asserts \`GCPServiceAttachmentAvailable = True\` — GCP enforces VPC consistency server-side when creating a Service Attachment, so a successful attachment is proof the NAT subnet and forwarding rule share the same VPC. No GCP credentials required in the test binary.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added end-to-end coverage for Google Cloud Private Service Connect (PSC) on HostedClusters, including validation of forwarding rule and NAT subnet fields plus service attachment availability.

* **Bug Fixes**
  * Improved PSC reconciliation to better discover missing forwarding rule and NAT subnet details, including correct NAT subnet selection when multiple candidates exist.
  * Ensures correct behavior when only one of the networking fields is set, preserving existing values.

* **Tests**
  * Expanded unit test coverage for PSC reconciliation logic, forwarding rule lookup behavior, NAT subnet filter formatting, and NAT subnet discovery edge cases.

### CODE REVIEW

File: hypershift-operator/controllers/platform/gcp/privateserviceconnect_controller_test.go [hunk 1]
```diff
@@ -291,3 +337,232 @@ func TestConstructServiceAttachmentName(t *testing.T) {
...
+	rule, err := r.lookupForwardingRule(context.Background(), newGCPPSC("", ""))
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+	if rule == nil || rule.Name != "fr-first" {
+		t.Errorf("expected first rule, got %v", rule)
+	}
+}
+
+// --- reconcileGCPPrivateServiceConnectSpec ---
```
Comment (cblecker) [inline hypershift-operator/controllers/platform/gcp/privateserviceconnect_controller_test.go:436]: Nit: the partial-write edge case (ForwardingRuleName set, NATSubnet empty) that the code comments on L211-213 call out is worth covering here. Something like `newGCPPSC("original-rule", "")` with a fake returning a *different* rule name would verify that the existing name is preserved and `discoverNATSubnet` still gets the correct network URL.

File: hypershift-operator/controllers/platform/gcp/privateserviceconnect_controller.go [hunk 2]
```diff
@@ -233,44 +292,38 @@ func (r *GCPPrivateServiceConnectReconciler) isSubnetInUse(ctx context.Context,
...
-				log.Error(err, "Failed to check subnet usage", "subnet", subnet.Name)
-				continue // Skip this subnet and try the next one
-			}
-
-			if !inUse {
-				log.Info("Found available PSC subnet", "subnet", subnet.Name)
-				return subnet.Name, nil
-			}
+	// Find the first available PSC subnet in the MC's VPC not already in use by another Service Attachment.
+	for _, subnet := range subnets {
```
Comment (cblecker) [inline hypershift-operator/controllers/platform/gcp/privateserviceconnect_controller.go:315]: Pre-existing, not introduced by this PR: `isSubnetInUse` makes a fresh `ListServiceAttachments` call per candidate subnet. In environments with several PSC subnets, fetching the list once before the loop and building a set of in-use names would avoid N redundant API calls. Fine as a follow-up.

Comment (cristianoveiga) [reply hypershift-operator/controllers/platform/gcp/privateserviceconnect_controller.go:315]: Created https://redhat.atlassian.net/browse/GCP-883 as a follow up.

File: hypershift-operator/controllers/platform/gcp/privateserviceconnect_controller_test.go (hunk 1 above)
Comment (cristianoveiga) [reply hypershift-operator/controllers/platform/gcp/privateserviceconnect_controller_test.go:436]: Addressed - added a new test case to cover this.

File: hypershift-operator/controllers/platform/gcp/privateserviceconnect_controller.go [hunk 3]
```diff
@@ -233,44 +292,38 @@ func (r *GCPPrivateServiceConnectReconciler) isSubnetInUse(ctx context.Context,
...
-			log.V(1).Info("Subnet already in use, trying next", "subnet", subnet.Name)
+		if !inUse {
+			log.Info("Found available PSC subnet", "subnet", subnet.Name)
+			return subnet.Name, nil
 		}
+
+		log.V(1).Info("Subnet already in use, trying next", "subnet", subnet.Name)
 	}
 
 	return "", fmt.Errorf("no available PRIVATE_SERVICE_CONNECT subnet found in region %s", r.Region)
```
Comment (cblecker) [inline hypershift-operator/controllers/platform/gcp/privateserviceconnect_controller.go:334]: Suggestion: when every `isSubnetInUse` call errors out (e.g. transient GCP IAM / API failure), this message reports "no available subnet" — which would send operators looking at subnet provisioning rather than API connectivity. Consider tracking error count through the loop and including it here, e.g. `"no available PRIVATE_SERVICE_CONNECT subnet found in region %s (failed to check %d of %d candidates due to API errors)"`.

**Comment by cblecker** [comment]
@coderabbit resume

**Review by cblecker — APPROVED** [review]

**Comment by cblecker** [comment]
Changes solely in GCP package. Azure confirmed infrastructure flake
/override ci/prow/e2e-azure-v2-self-managed

/verified later by @cristianoveiga
